uint8[] groups
---
string state
